src/object-store/src/layers/lru_cache/read_cache.rs

```diff
@@ -0,0 +1,290 @@
+    file_cache: Arc<C>,
+    /// Local memory cache to track local cache files
+    mem_cache: Cache<String, ReadResult>,
+}
+
+impl<C: Accessor + Clone> ReadCache<C> {
+    /// Create a [`ReadCache`] with capacity in bytes.
+    pub(crate) fn new(file_cache: Arc<C>, capacity: usize) -> Self {
+        let file_cache_cloned = file_cache.clone();
+        let eviction_listener =
+            move |read_key: Arc<String>, read_result: ReadResult, cause| -> ListenerFuture {
+                // Delete the file from local file cache when it's purged from mem_cache.
+                decrement_gauge!(OBJECT_STORE_LRU_CACHE_ENTRIES, 1.0);
+                let file_cache_cloned = file_cache_cloned.clone();
+
+                async move {
+                    if let ReadResult::Success(size) = read_result {
+                        decrement_gauge!(OBJECT_STORE_LRU_CACHE_BYTES, size as f64);
+
+                        let result = file_cache_cloned.delete(&read_key, OpDelete::new()).await;
+                        debug!(
+                            "Deleted local cache file `{}`, result: {:?}, cause: {:?}.",
+                            read_key, result, cause
+                        );
+                    }
+                }
+                .boxed()
+            };
+
+        Self {
+            file_cache,
+            mem_cache: Cache::builder()
+                .max_capacity(capacity as u64)
+                .weigher(|_key, value: &ReadResult| -> u32 {
+                    // TODO(dennis): add key's length to weight?
+                    value.size_bytes()
+                })
+                .async_eviction_listener(eviction_listener)
+                .support_invalidation_closures()
+                .build(),
+        }
+    }
+
+    /// Returns the cache's entry count and total approximate entry size in bytes.
+    pub(crate) async fn stat(&self) -> (u64, u64) {
+        self.mem_cache.run_pending_tasks().await;
+
+        (self.mem_cache.entry_count(), self.mem_cache.weighted_size())
+    }
+
+    /// Invalidte all cache items which key starts with `prefix`.
+    pub(crate) async fn invalidate_entries_with_prefix(&self, prefix: String) {
+        // Safety: always ok when building cache with `support_invalidation_closures`.
+        self.mem_cache
+            .invalidate_entries_if(move |k: &String, &_v| k.starts_with(&prefix))
+            .ok();
+    }
+
+    /// Blocking version of `invalidate_entries_with_prefix`.
+    pub(crate) fn blocking_invalidate_entries_with_prefix(&self, prefix: String) {
+        // Safety: always ok when building cache with `support_invalidation_closures`.
+        self.mem_cache
+            .invalidate_entries_if(move |k: &String, &_v| k.starts_with(&prefix))
+            .ok();
+    }
+
+    /// Recover existing cache items from `file_cache` to `mem_cache`.
+    /// Return entry count and total approximate entry size in bytes.
+    pub(crate) async fn recover_cache(&self) -> Result<(u64, u64)> {
```

Comment:
nitpick(non-blocking): Maybe we could add some debug or info logs before/after `recover_cache`.
```rust
debug!("Recover cache begin, path: {}", self.file_cache.info().root());

// ...

let stat = self.stat().await;
debug!("Recover cache end, entry count: {}, entry size: {}", stat.0, stat.1);
```

Reply:
Already logged in `LruCacheLayer`.

```rust
        info!(
            "Recovered {} entries and total size {} in bytes for LruCacheLayer",
            entries, bytes
        );

```